src/nbl/asset/ICPUImage.cpp

```diff
@@ -0,0 +1,319 @@
+
+			scratch.hashes = { .offset = 0u, .size = product * sizeof(CState::outHash), .buffer = buffer };
+			scratch.hashers = { .offset = scratch.hashes.size, .size = product * sizeof(blake3_hasher), .buffer = buffer };
+			scratch.flatten = { .offset = scratch.hashers.offset + scratch.hashers.size, .size = state->scratch.size - scratch.hashers.size - scratch.hashes.size, .buffer = buffer };
+		}
+
+		const auto isFullyFlatten = scratch.flatten.size == 0ull;
+
+		struct
+		{
+			CFlattenRegionsImageFilter::state_type flatten;	// CFlattenRegionsImageFilter's state, we use scratch memory to respecify regions & prefill with 0-value non-covered texels
+		} proxy;
+
+		/*
+			first we need to ensure that texels which are not covered by regions
+			are filled with 0 value, we use flatten with tight buffer for the
+			texel copy & prefill if we don't have fully flatten input
+		*/
+
+		if (!isFullyFlatten) // TODO: we may think of even better optimization if we have mixed regions for mips (eg. a few flattened, some others not and left empty)
+		{
+			// create own regions & hook tight buffer with no gaps from scratch
+			auto regions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<IImage::SBufferCopy>>(parameters.mipLevels);
+			{
+				size_t bufferSize = 0ull;
+
+				for (auto rit = regions->begin(); rit != regions->end(); rit++)
+				{
+					auto miplevel = static_cast<uint32_t>(std::distance(regions->begin(), rit));
+					auto localExtent = state->inImage->getMipSize(miplevel);
+					rit->bufferOffset = bufferSize;
+					rit->bufferRowLength = localExtent.x; // could round up to multiple of 8 bytes in the future
+					rit->bufferImageHeight = localExtent.y;
+					rit->imageSubresource.aspectMask = IImage::E_ASPECT_FLAGS::EAF_COLOR_BIT; // otherwise won't pass validaiton 
+					rit->imageSubresource.mipLevel = miplevel;
+					rit->imageSubresource.baseArrayLayer = 0u;
+					rit->imageSubresource.layerCount = parameters.arrayLayers;
+					rit->imageOffset = { 0u,0u,0u };
+					rit->imageExtent = { localExtent.x,localExtent.y,localExtent.z };
+					auto levelSize = info.roundToBlockSize(localExtent);
+					auto memsize = levelSize[0] * levelSize[1] * levelSize[2] * parameters.arrayLayers * bytesPerPixel;
+
+					assert(memsize.getNumerator() % memsize.getDenominator() == 0u);
+					bufferSize += memsize.getIntegerApprox();
+				}
+			}
+
+			auto out = ICPUImage::create(IImage::SCreationParams(parameters));
+			out->setBufferAndRegions(core::smart_refctd_ptr(getScratchAsBuffer(scratch.flatten.size, scratch.flatten.offset)), std::move(regions));
+
+			auto& flatten = proxy.flatten;
+			flatten.inImage = state->inImage;
+			flatten.outImage = out;
+			flatten.preFill = true;
+			memset(flatten.fillValue.pointer, 0, sizeof(flatten.fillValue.pointer));
+
+			assert(CFlattenRegionsImageFilter::execute(policy, &proxy.flatten)); // this should never fail, at this point we are already validated
```

Comment:
@AnastaZIuk @tad1 I don't know if you already came across this, but any code in an `assert(blah)` just disappears in a non-Debug build (cause assert is a macro thats empty in a Release build)